Push to hub in a pull request

AJDERS opened the pull request:
Fixes #5492.

Introduce new kwarg `create_pr` in `push_to_hub`, which is passed to `HFapi.upload_file`.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/datasets/pr_5528). All of your documentation changes will be reflected on that endpoint.

albertvillanova reviewed (changes requested):
Thanks for the enhancement, @AJDERS.

Please note that the parameter `create_pr` is present in `huggingface_hub` only from huggingface_hub-0.9.0, whereas at `datasets` we support older versions (from 0.2.0).

To fix this we should check the installed version of huggingface_hub. You can have a look at the file `utils/_hf_hub_fixes` and maybe implement there the `upload_file` function, similarly to the other functions defined there.

AJDERS commented:
It seems that the parameter `create_pr` is available for [`0.8.0`](https://huggingface.co/docs/huggingface_hub/v0.8.1/en/package_reference/hf_api#huggingface_hub.HfApi.upload_file) (its not here: [`0.7.0`](https://huggingface.co/docs/huggingface_hub/v0.7.0.rc0/en/package_reference/hf_api#huggingface_hub.HfApi.upload_file)) and onwards. I included a warning, informing the user that no PR was created.

nateraw reviewed (commented):
do you ever set `create_pr` to False while iterating over shards? If not, you'll create a new PR for each shard if I'm not mistaken.

Below, I've pseudocoded out an idea of the logic (this won't run).

```python
from huggingface_hub import create_pull_request, upload_file, create_repo, delete_repo, get_repo_discussions

revision = 'refs/pr/1'
create_pr = True
repo_name = 'test-pr-upload-file-datasets'

repo_url = create_repo(repo_name, exist_ok=True, repo_type='dataset')
repo_id = repo_url.repo_id

# Try to find PR revision if its supplied on push_to_hub
revision_found = False
if create_pr is True and revision is not None:
    for discussion in get_repo_discussions(repo_id, repo_type='dataset'):
        if discussion.is_pull_request and discussion.git_reference == revision:
            create_pr = False
            revision_found = True
            break
    if not revision_found:
        raise ValueError("Provided revision not found")

# If no revision provided, create the PR and store the revision
if create_pr:
    pr = create_pull_request(repo_id, repo_type='dataset')
    revision = pr.git_reference
    create_pr = False

# Lots of logic that leads to this basically...
shards = ['test-shard-0.parquet', 'test-shard-1.parquet', 'test-shard-2.parquet']

# Upload each shard to the hub @ the specified revision
for shard in shards:
    upload_file(
        repo_id,
        repo_type='dataset',
        path_or_fileobj=shard,
        path_in_repo=f'data/test/{shard}',
        revision=revision,
    )
```

CC @Wauplin for advice here.

AJDERS commented:
@nateraw you are completely right! Actually, the dataset shards is never added to the created pr, only the metadata, as the code is now. Ill look into you suggestion asap. Thank!

Wauplin commented:
@nateraw Nothing more to add, that's a perfect usage of `huggingface_hub` as far as I can tell! :)

A very nit improvement would be to use the [for .. else ... python statement](https://book.pythontips.com/en/latest/for_-_else.html).
i.e:

```py
if create_pr is True and revision is not None:
    for discussion in get_repo_discussions(repo_id, repo_type='dataset'):
        if discussion.is_pull_request and discussion.git_reference == revision:
            create_pr = False
            break
    else:
        raise ValueError("Provided revision not found")
```
No need for the `revision_found` temporary flag when do so. Yeah ok, it's niche :wink:

AJDERS commented:
I added the suggestions from @nateraw and @Wauplin

albertvillanova commented on src/datasets/utils/_hf_hub_fixes.py:
```diff
@@ -172,3 +176,46 @@ def list_repo_files(
+    Several new parameters for huggingface_hub.HfApi.upload_file were introduced in 0.8.1 and some of them were deprecated.
+    """
+    if version.parse(huggingface_hub.__version__) < version.parse("0.8.1"):
+        logger.warning(
+            "The `create_pr` parameter is only available for huggingface_hub>=0.8.0. Hence no pull request is created. Please upgrade huggingface_hub to >=0.8.0 to create a pull request."
+        )
+        return hf_api.upload_file(
+            path_or_fileobj=path_or_fileobj,
+            path_in_repo=path_in_repo,
+            repo_id=repo_id,
+            token=token,
+            repo_type=repo_type,
+            revision=revision,
+        )
```
I'm thinking that maybe we could raise a TypeError in this case, so that we avoid creating a commit to main branch when the user intended to open a PR. What do you think?

The error message should be similar to the one you created for the warning. The user should either:
- update huggingface_hub to >=0.8 if they want to create a pull request (please note that 0.8.0 version was yanked: https://pypi.org/project/huggingface-hub/#history)
- not pass `create_pr` if they do not want to create a PR

AJDERS replied:
I agree, and i have added it below.

albertvillanova reviewed (changes requested):
Thanks. Some comments/suggestions below...

Why have you removed the test for create_pr? You could add it again and just add a pytest skipif when version of huggingface_hub is lower than 0.8.1.

albertvillanova commented on src/datasets/utils/_hf_hub_fixes.py:
```diff
@@ -172,3 +176,46 @@ def list_repo_files(
+    """
+    Several new parameters for huggingface_hub.HfApi.upload_file were introduced in 0.8.1 and some of them were deprecated.
+    """
+    if version.parse(huggingface_hub.__version__) < version.parse("0.8.1"):
```
```suggestion
    if create_pr is not None and version.parse(huggingface_hub.__version__) < version.parse("0.8.1"):
```

albertvillanova reviewed (commented):
Also, the commit history of this PR is full of commits from other PRs...

I don't know how you incorporated them. But for the future, please note that you can cleanly incorporate the changes from the main branch with a merge commit:
```
git checkout push-to-hub-in-a-pull-request
git fetch upstream main
git merge upstream/main
git push origin push-to-hub-in-a-pull-request
```

AJDERS commented:
> Thanks. Some comments/suggestions below...
> 
> Why have you removed the test for create_pr? You could add it again and just add a pytest skipif when version of huggingface_hub is lower than 0.8.1.

I have added the test again. I removed it because i kept getting errors when calling `create_pull_request` with `repo_id=ds_name` where `temporary_repo = ds_name`, and thought i might look more thoroughly at it later. I have added a test called `test_test` showing this, it gives:
```
tests/test_upstream_hub.py:360: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.venv/lib/python3.10/site-packages/huggingface_hub/utils/_validators.py:124: in _inner_fn
    return fn(*args, **kwargs)
.venv/lib/python3.10/site-packages/huggingface_hub/hf_api.py:3451: in create_pull_request
    return self.create_discussion(
.venv/lib/python3.10/site-packages/huggingface_hub/utils/_validators.py:124: in _inner_fn
    return fn(*args, **kwargs)
.venv/lib/python3.10/site-packages/huggingface_hub/hf_api.py:3393: in create_discussion
    hf_raise_for_status(resp)
(...)
E               huggingface_hub.utils._errors.RepositoryNotFoundError: 401 Client Error. (Request ID: Root=1-63ecd2cb-2cf2557a332c86ad27f687b3)
E               
E               Repository Not Found for url: https://huggingface.co/api/models/__DUMMY_TRANSFORMERS_USER__/test-16764648321590/discussions.
E               Please make sure you specified the correct `repo_id` and `repo_type`.
E               If you are trying to access a private or gated repo, make sure you are authenticated.
E               Invalid username or password.
```

Wauplin commented on src/datasets/arrow_dataset.py:
```diff
@@ -5138,6 +5148,34 @@ def push_to_hub(
+        try:
+            api.repo_info(repo_id, token=token)
+        except RepositoryNotFoundError:
+            repo_url = create_repo(hf_api=api, repo_id=repo_id, token=token, exist_ok=True, repo_type="dataset")
+            repo_id = repo_url.repo_id
```
This is the correct way of getting the `repo_id` when creating a repository (including potentially resolving the namespace if not explicitly provided) _BUT_ it is only available since `huggingface_hub>=0.12`. Before that, the return value was a pure `str` representing the URL (and has to be manually parsed). I don't know how it was handled in `datasets` before, maybe worth to keep it consistent (e.g. most probably keeping the `repo_id` provided by the user and assume it is the correct one). 

For the record, the new return value is of type `RepoUrl` which provides info like endpoint, repo_type, namespace and repo_id. `RepoUrl` is a subclass of `str` so it doesn't bring any breaking changes.

AJDERS replied:
I checked 2.7, 2.6 and 1.18.X for how the output of `create_repo` was parsed, and it does not seem the output is ever used, hence it does not seem there is preferred way to parse the URL? The output is simply `https://huggingface.co/{repo_id}`, so maybe:
``` python
from urllib.parse import urlparse
repo_id = urlparse(repo_url).path[:1]
```

lhoestq replied:
Maybe you can move this logic in `_hf_hub_fixes` as well ?

AJDERS replied:
Like [this](https://github.com/huggingface/datasets/pull/5528/commits/12327e310b2086a5cb0e51c23677ba6a3102a55a) ?

Wauplin commented on src/datasets/arrow_dataset.py:
```diff
@@ -5138,6 +5148,34 @@ def push_to_hub(
+
+        # Try to find PR branch if branch is supplied
+        if create_pr and branch is not None:
+            for discussion in get_repo_discussions(repo_id, repo_type="dataset"):
```
Actually, this has been introduced in [huggingface 0.9 release](https://github.com/huggingface/huggingface_hub/releases/tag/v0.9.0). @albertvillanova  @lhoestq how do you deal with stuff like this ?

AJDERS replied:
For the `get_repo_discussions`-problem, maybe do like with `upload_file` in `utils._hf_hub_fixes.py`, i.e.:
```
def get_repo_discussions(
    hf_api: HfApi,
    repo_id: str,
    repo_type: Optional[str] = None,
    token: Optional[str] = None,
) -> Iterator[Discussion]:
    """
    The method `huggingface_hub.HfApi.get_repo_discussions` was introduced in 0.9.0.
    """
    if version.parse(huggingface_hub.__version__) < version.parse("0.9.0"):
        raise TypeError(
            "The method `get_repo_discussions` was introduced in huggingface_hub 0.9.0. Please update huggingface_hub to >=0.9.0 to use this method."
        )
    else:
        return hf_api.get_repo_discussions(
            repo_id=repo_id,
            repo_type=repo_type,
            token=token,
        )
```

lhoestq replied:
Hi ! I think you can check for `huggingface_hub` version directly in `push_to_hub` and provide a more specialized error message:
```python
"Using `create_pr` requires `huggingface-hub>=0.9.0`. Please use a more recent version of `huggingface-hub`."
```

AJDERS replied:
Like [this](https://github.com/huggingface/datasets/pull/5528/commits/c7a998f28135c1a68d6e603fb036c5776225c6ab) ?

albertvillanova replied:
@lhoestq I remember we decided to put all huggingface_hub version-dependence logic in one single place (file `_hf_hub_fixes.py`). The reason was to make things easier once we decide to stop supporting older hfh versions...

lhoestq replied:
Yea I guess it'b be nice to still have the function in `_hf_hub_fixes`. But having the specialized error is necessary to not be confusing for users

albertvillanova replied:
@lhoestq: https://github.com/huggingface/datasets/pull/4106#issuecomment-1090283193

> adrinjalali: I find it good practice to have all dependency version related code in a single file so that when you decide to remove support for an old version of a dependency it's easy to find and remove them

> lhoestq: good idea, thanks !

> Commit: [move version specific code to its own file](https://github.com/huggingface/datasets/pull/4106/commits/399aaf47113eed93d9b786e822064584505969c3)

AJDERS replied:
Ill move the function back into `_hf_hub_fixes` and leave the error message.

Wauplin reviewed (commented):
PR looks good to me from a logic perspective.

I left 2 comments things that would not work with old versions of `huggingface_hub`. I'll let @albertvillanova @lhoestq suggest how we could change this (raise a `TypeError` I guess?)

lhoestq commented on src/datasets/arrow_dataset.py:
```diff
@@ -5138,6 +5150,39 @@ def push_to_hub(
+                    "Using `create_pr` requires `huggingface-hub>=0.9.0`. Please use a more recent version of `huggingface-hub`."
+                )
+            else:
+                for discussion in get_repo_discussions(repo_id, repo_type="dataset"):
```
You probably need to import `get_repo_discussions` here. At the top of the file it may fail

lhoestq reviewed (commented):
Yes thanks ! Feel free to run `make style` to fix the code formatting

AJDERS commented:
> > Thanks. Some comments/suggestions below...
> > Why have you removed the test for create_pr? You could add it again and just add a pytest skipif when version of huggingface_hub is lower than 0.8.1.
> 
> I have added the test again. I removed it because i kept getting errors when calling `create_pull_request` with `repo_id=ds_name` where `temporary_repo = ds_name`, and thought i might look more thoroughly at it later. I have added a test called `test_test` showing this, it gives:
> 
> ```
> tests/test_upstream_hub.py:360: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> .venv/lib/python3.10/site-packages/huggingface_hub/utils/_validators.py:124: in _inner_fn
>     return fn(*args, **kwargs)
> .venv/lib/python3.10/site-packages/huggingface_hub/hf_api.py:3451: in create_pull_request
>     return self.create_discussion(
> .venv/lib/python3.10/site-packages/huggingface_hub/utils/_validators.py:124: in _inner_fn
>     return fn(*args, **kwargs)
> .venv/lib/python3.10/site-packages/huggingface_hub/hf_api.py:3393: in create_discussion
>     hf_raise_for_status(resp)
> (...)
> E               huggingface_hub.utils._errors.RepositoryNotFoundError: 401 Client Error. (Request ID: Root=1-63ecd2cb-2cf2557a332c86ad27f687b3)
> E               
> E               Repository Not Found for url: https://huggingface.co/api/models/__DUMMY_TRANSFORMERS_USER__/test-16764648321590/discussions.
> E               Please make sure you specified the correct `repo_id` and `repo_type`.
> E               If you are trying to access a private or gated repo, make sure you are authenticated.
> E               Invalid username or password.
> ```

@albertvillanova FYI I have looked at this again, and i haven't figured it out, so the test`test_push_dataset_to_hub_with_pull_request` and the minimal example `test_test` are still failing locally.

albertvillanova reviewed (commented):
For the tests, I am having a look at your code...

albertvillanova reviewed (changes requested):
Please fix the style by running:
```
pip install "black~=23.1" "ruff>=0.0.241"
make style
git add -u
git commit -m "Fix style"
git push origin push-to-hub-in-a-pull-request
```

albertvillanova commented on src/datasets/utils/_hf_hub_fixes.py:
```diff
@@ -1,10 +1,17 @@
 from huggingface_hub import HfApi, HfFolder
-from huggingface_hub.hf_api import DatasetInfo
+from huggingface_hub.community import Discussion
+from huggingface_hub.hf_api import DatasetInfo, RepoUrl
```
This import gives an error:
```
E   ImportError: cannot import name 'RepoUrl' from 'huggingface_hub.hf_api'
```

albertvillanova commented on src/datasets/arrow_dataset.py:
```diff
@@ -52,12 +52,15 @@
 import pyarrow as pa
 import pyarrow.compute as pc
 from huggingface_hub import HfApi, HfFolder
+from huggingface_hub.utils import RepositoryNotFoundError
```
This gives an error:
```
E   ImportError: cannot import name 'RepositoryNotFoundError' from 'huggingface_hub.utils'
```

albertvillanova reviewed (changes requested):
Just while trying to run the tests, I have got some import errors.

I guess we should try to find if there were changes across `huggingface_hub` versions and fix those with local imports in `_hf_hub_fixes.py`.

AJDERS commented:
I tried to move all of the "create pr safely"-logic to a seperate function in `_hf_hub_fixes`. I looked at how the exceptions were raised before `huggingface_hub.utils.RepositoryNotFoundError`existed, and make changes accordingly.

albertvillanova commented on src/datasets/utils/_hf_hub_fixes.py:
```diff
@@ -60,6 +66,111 @@ def create_repo(
+
+    # Try to find the repo before creating it
+    try:
+        hf_api.repo_info(repo_id, token=token)
```
Missing repo_type:
```suggestion
        hf_api.repo_info(repo_id, repo_type=repo_type, token=token)
```

albertvillanova commented on src/datasets/utils/_hf_hub_fixes.py:
```diff
@@ -60,6 +66,111 @@ def create_repo(
+    if version.parse(huggingface_hub.__version__) < version.parse("0.12.0"):
+        from urllib.parse import urlparse
+
+        repo_id = urlparse(repo_url).path[:1]
```
This code line is wrong:
- for repo_url = "https://hub-ci.huggingface.co/datasets/__DUMMY_TRANSFORMERS_USER__/test-16766285471927"
- it gives repo_id = "/"

albertvillanova commented on src/datasets/utils/_hf_hub_fixes.py:
```diff
@@ -60,6 +66,111 @@ def create_repo(
+        repo_url = create_repo(
+            hf_api=hf_api, repo_id=repo_id, private=private, token=token, exist_ok=True, repo_type=repo_type
+        )
+        repo_id = get_repo_id_from_repo_url(repo_url)
```
I would remove this line because we already have `repo_id` (we just passed it as parameter to the function call above)
```suggestion
```

albertvillanova commented on src/datasets/utils/_hf_hub_fixes.py:
```diff
@@ -60,6 +66,111 @@ def create_repo(
+
+    # Create PR if we didn't find it before
+    if create_pr:
+        from huggingface_hub import create_pull_request
+
+        pr = create_pull_request(
```
Wrong API:
```suggestion
        pr = hf_api.create_pull_request(
```

albertvillanova commented on src/datasets/utils/_hf_hub_fixes.py:
```diff
@@ -60,6 +66,111 @@ def create_repo(
+
+        pr = create_pull_request(
+            repo_id,
+            repo_type="dataset",
```
Nit:
```suggestion
            repo_type=repo_type,
```

albertvillanova reviewed (changes requested):
Thanks again. This is turning out to be quite a complex pull request, because besides the addition of the `create_pr` parameter, this also adds more logic: create repo, look for existing PR branch, create PR separately,... Maybe we should have split this into several different pull requests...

I have found some errors... see below...

albertvillanova commented on src/datasets/utils/_hf_hub_fixes.py:
```diff
@@ -60,6 +66,111 @@ def create_repo(
+            token=token,
+        )
+        branch = pr.git_reference
+        create_pr = False
```
This value is never used...

albertvillanova commented on src/datasets/utils/_hf_hub_fixes.py:
```diff
@@ -60,6 +66,112 @@ def create_repo(
+    if version.parse(huggingface_hub.__version__) < version.parse("0.12.0"):
+        from urllib.parse import urlparse
+
+        repo_id = urlparse(repo_url).path[1:]
```
This line is still wrong for repo_type = "dataset" (it is only valid for repo_type=None or "model"):
- for repo_url = "https://hub-ci.huggingface.co/datasets/__DUMMY_TRANSFORMERS_USER__/test-16766285471927"
- it gives repo_id = 'datasets/__DUMMY_TRANSFORMERS_USER__/test-16766285471927'
- instead of: '__DUMMY_TRANSFORMERS_USER__/test-16766285471927'

albertvillanova commented on tests/test_upstream_hub.py:
```diff
@@ -331,6 +336,37 @@ def test_push_dataset_to_hub(self, temporary_repo):
+                assert list(local_ds.features.keys()) == list(hub_ds.features.keys())
+                assert local_ds.features == hub_ds.features
+
+    def test_test(self, temporary_repo):
+        local_ds = Dataset.from_dict({"x": [1, 2, 3], "y": [4, 5, 6]})
+        with temporary_repo(f"{CI_HUB_USER}/test-{int(time.time() * 10e3)}") as ds_name:
+            local_ds.push_to_hub(ds_name, split="train", token=self._token)
+            huggingface_hub.create_pull_request(
+                repo_id=ds_name,
+                title="Test PR",
+                token=self._token,
+                repo_type="dataset",
+            )
+
```
I guess you can remove this test. Indeed you are using the wrong API in `huggingface_hub.create_pull_request`

albertvillanova reviewed (changes requested):
Just some minor fixes to have the CI green...

AJDERS commented:
`create_pr` was set during `push_to_hub`, even though it was `None` from the outset, hence causing tests to fail for older versions of `huggingface_hub`. This is now fixed.

With the implementation of `_hf_hub_fixes.upload_file` the function call expected `commit_message`, `commit_description`. If these are not set we call the function without them, even though we are on a version of `huggingface_hub` where they are not available in `upload_file`.

Many tests failed for `dataset_dict` for the above reasons, so the fixes from `arrow_dataset.py` were also added to `dataset_dict.py`.

AJDERS commented:
`create_pr` was set during `push_to_hub`, even though it was `None` from the outset, hence causing tests to fail for older versions of `huggingface_hub`. This is now fixed.

With the implementation of `_hf_hub_fixes.upload_file` the function call expected `commit_message`, `commit_description`. If these are not set we call the function without them, even though we are on a version of `huggingface_hub` where they are not available in `upload_file`.

When `huggingface_hub < 0.5.0` we assume `repo_id` of them form `organisation/name`, so now that we are calling `create_repo` in the tests with `repo_id` not of this form, we need to handle this case, this is now done.

Many tests failed for `dataset_dict` for the above reasons, so the fixes from `arrow_dataset.py` were also added to `dataset_dict.py`. 

**All tests are now passing locally for `huggingface_hub==0.2.0` and `huggingface_hub==0.12.1`…** Im sorry I should have downgraded and went through this a long time ago, but I didn’t realise the extend of these version fixes until recently…

lhoestq commented:
Hi ! FYI bumped the `huggingface-hub` dependency to 0.11 and removed the `_hf_hub_fixes.py` - which should make this PR much easier

abzdel commented:
Just now finding this - seems like a cool issue to contribute to. If any more help is needed please ping me! @AJDERS
